src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -281,19 +296,46 @@ private class RsFnInferenceContext(
 
     private fun RsExpr.inferTypeCoercableTo(expected: Ty): Ty {
         val inferred = inferType(expected)
-        coerce(inferred, expected)
+        coerce(this, inferred, expected)
         return inferred
     }
 
-    private fun coerce(inferred: Ty, expected: Ty) {
+    private fun coerce(expr: RsExpr, inferred: Ty, expected: Ty) {
+        val ok = tryCoerce(inferred, expected)
+        if (!ok) {
+            val expected = ctx.resolveTypeVarsIfPossible(expected)
+            val inferred = ctx.resolveTypeVarsIfPossible(inferred)
+            // ignoring possible false-positives (it's only basic experimental type checking)
+            val ignoredTys = listOf(
+                TyUnknown::class.java,
+                TyInfer::class.java,
+                TyTypeParameter::class.java,
+                // TODO TyReference ignored because we actually ignore deref level on method call and so
+                // TODO sometimes substitute a wrong receiver. This should be fixed as soon as possible
+                TyReference::class.java,
+                TyTraitObject::class.java
+            )
+
+            if (!expected.containsTyOfClass(ignoredTys) && !inferred.containsTyOfClass(ignoredTys)) {
+                // another awful hack: check that inner expressions did not annotated as an error
+                // to disallow annotation intersections. This should be done in a different way
+                fun PsiElement.isChildOf(psi: PsiElement) = this.ancestors.contains(psi)
+                if (ctx.diagnostics.all { !it.element.isChildOf(expr) }) {
+                    ctx.reportTypeMismatch(expr, expected, inferred)
+                }
```

Comment:
This is probably OK for now, but it's definitely superlinear in complexity. 

I think a better architecture here would be to collect **all** diagnostics at first and the filter out some of them at the very end.